File: src/rust/planning_poker/src/usecases/change_user_name.rs
use crate::{
    domains::user::{HaveUserRepository, User, UserId, UserRepository},
    utils::types::LocalBoxFuture,
};

#[cfg(test)]
mod tests;

pub trait ChangeUserNameDependency: HaveUserRepository {}

#[derive(PartialEq, Debug)]
pub enum ChangeUserNameOutput {
    NotFound,
    CanNotChangeName,
}

pub trait ChangeUserName {
    fn execute(
        &self,
        user_id: UserId,
        name: &str,
    ) -> LocalBoxFuture<'_, Result<User, ChangeUserNameOutput>>;
}

impl<T: ChangeUserNameDependency> ChangeUserName for T {
    fn execute<'a>(
        &'a self,
        user_id: UserId,
        name: &str,
    ) -> LocalBoxFuture<'a, Result<User, ChangeUserNameOutput>> {
        let repository = self.get_user_repository();
        let name = name.to_owned();

        let fut = async move {
            let user = repository.find_by(user_id).await;

            match user {
                None => Err(ChangeUserNameOutput::NotFound),
                Some(mut user) => {
                    if !User::can_change_name(&name) {
                        return Err(ChangeUserNameOutput::CanNotChangeName);
                    }

                    user.change_name(&name);
                    repository.save(&user).await;

                    Ok(user)
                }
            }
        };

        Box::pin(fut)
    }
}

File: src/usecases/change-user-name.ts
import * as User from "@/domains/user";
import { UserRepository } from "@/domains/user-repository";
import { EventDispatcher, UseCase } from "./base";

export interface ChangeUserNameInput {
  userId: User.Id;
  name: string;
}

export type ChangeUserNameOutput =
  | { kind: "success"; user: User.T }
  | { kind: "canNotChangeName" }
  | { kind: "notFound" };

export class ChangeUserNameUseCase implements UseCase<ChangeUserNameInput, Promise<ChangeUserNameOutput>> {
  constructor(private dispatcher: EventDispatcher, private userRepository: UserRepository) {}

  async execute(input: ChangeUserNameInput): Promise<ChangeUserNameOutput> {
    const user = await this.userRepository.findBy(input.userId);
    if (!user) {
      return { kind: "notFound" };
    }

    if (!User.canChangeName(input.name)) {
      return { kind: "canNotChangeName" };
    }

    const [newUser, event] = User.changeName(user, input.name);
    this.userRepository.save(newUser);

    if (event) {
      this.dispatcher.dispatch(event);
    }

    return { kind: "success", user: newUser };
  }
}
